src/fsharp/service/service.fs

```diff
@@ -443,20 +444,19 @@ type BackgroundCompiler(legacyReferenceResolver, projectCacheSize, keepAssemblyC
                 return FSharpParseFileResults(diagnostics = diagnostics, input = parseTree, parseHadErrors = false, dependencyFiles = builder.AllDependenciesDeprecated)
         }
 
-    member _.GetCachedCheckFileResult(builder: IncrementalBuilder, filename, sourceText: ISourceText, options) =
-        // Check the cache. We can only use cached results when there is no work to do to bring the background builder up-to-date
-        let cachedResults = parseCacheLock.AcquireLock (fun ltok -> checkFileInProjectCache.TryGet(ltok, (filename, sourceText.GetHashCode(), options)))
-
-        match cachedResults with 
-        | Some (parseResults, checkResults,_,priorTimeStamp) 
-                when 
-                (match builder.GetCheckResultsBeforeFileInProjectEvenIfStale filename with 
-                | None -> false
-                | Some(tcPrior) -> 
-                    tcPrior.TimeStamp = priorTimeStamp &&
-                    builder.AreCheckResultsBeforeFileInProjectReady(filename)) -> 
-            Some (parseResults,checkResults)
-        | _ -> None
+    member _.GetCachedCheckFileResult(builder: IncrementalBuilder, filename, sourceText: ISourceText, options, cacheStamp: int64 option) =
+        // If a cache stamp is not provided, we need to do a full-up-to-date check on the timestamps for the dependencies.
+        let recheckDeps = cacheStamp.IsNone
+        if recheckDeps && not (builder.AreCheckResultsBeforeFileInProjectReady filename) then None
+        else
+            let cacheStamp = defaultArg cacheStamp (sourceText.GetHashCode() |> int64)
+
+            // Check the cache. We can only use cached results when there is no work to do to bring the background builder up-to-date
+            let cachedResults = parseCacheLock.AcquireLock (fun ltok -> checkFileInProjectCache.TryGet(ltok, (filename, cacheStamp, options)))
```

Comment:
@TIHan This check doesn't work properly for me: changes in prior files don't make subsequent files get reanalyzed when `cacheSame` is not used. Consider the following repro steps for project with two files:
* Parse and check File2.fs (File1.fs is checked as a dependency, results for File1 and File2 get cached)
* Change File1.fs and parse and check it
* Parse and check File2.fs again, it'll have the same stamp (since its source hasn't changed)

Reply:
The `AreCheckResultsBeforeFileInProjectReady` should be returning `false` in your case, perhaps I'll force clear the cache since it's technically invalid at that point.

Reply:
Should we able to use `CheckFileInProjectAllowingStaleCachedResults` for invalid results? It might be OK for some features.

Reply:
I think that's fair, but I don't know if FCS should dictate when to use stale cached results as each editor will be different.

Reply:
`CheckFileInProjectAllowingStaleCachedResults` looks like it's considered Obsolete and encourages us to use `CheckFileInProject`. I think we should consider not making it Obsolete and just have a separate call to get cached results.